File: sdk/typescript/src/evaluator.ts
/**
 * Local invariant evaluator — mirrors the server-side `core::policy`
 * runtime checks in TypeScript so the SDK can DENY without an HTTP
 * roundtrip per tool invocation. Semantics MUST stay byte-equivalent
 * with the Rust evaluator; see `core/src/policy/invariants/*.rs`.
 *
 * @module
 */

import type { CompiledPolicy } from "./policy-cache";

/** One tool invocation to evaluate. Mirrors the server `Action` struct. */
export interface Action {
    /** Caller-supplied unique id (also used in receipts). */
    actionId: string;
    /** Tool/method to call (e.g. `http_get`, `sepa_payment_initiate`). */
    tool: string;
    /** USD amount if the action moves money. */
    amountUsd?: number;
    /** Data classification tag of the resource touched. */
    dataClassification?: string;
    /** Roles that have signed this action. */
    signatures: string[];
    /** How many delegation hops separate this action from the root agent. */
    delegationDepth: number;
    /** Unix-epoch seconds when the action was created. */
    timestamp: number;
}

/** Read-only context for one evaluation. */
export interface EvaluationContext {
    /** Cumulative USD spend so far. */
    spendTotalUsd: number;
    /** Unix-epoch *seconds* of recent calls (rate-limit input). */
    recentCallTimestamps: number[];
    /** Current unix-epoch seconds. */
    nowEpoch: number;
    /** `HH:MM` 24-hour in the policy's timezone. */
    nowTzHhmm: string;
}

/** Allow / deny result of one evaluation. */
export type Verdict =
    | { kind: "allow" }
    | { kind: "deny"; check: string; reason: string };

const RATE_WINDOW_SECS = 60;

/**
 * Run every applicable check from `policy.binding` against `action`.
 * Returns the first deny verdict, or `{ kind: "allow" }` if all pass.
 *
 * Order mirrors `core::policy::compiler::compile`:
 *   allowlist → budget → scope → rate_limit → time_window → signatures → delegation_depth.
 */
export function evaluate(
    policy: CompiledPolicy,
    action: Action,
    ctx: EvaluationContext
): Verdict {
    const b = policy.binding;

    // 1. allowlist (tool name)
    if (b.allowed_tools) {
        if (!b.allowed_tools.includes(action.tool)) {
            return {
                kind: "deny",
                check: "allowlist",
                reason: `tool '${action.tool}' not in allowlist`,
            };
        }
    }

    // 2. budget
    if (typeof b.max_budget_usd === "number") {
        const amount = action.amountUsd ?? 0;
        const projected = ctx.spendTotalUsd + amount;
        if (projected > b.max_budget_usd) {
            return {
                kind: "deny",
                check: "budget",
                reason: `projected spend ${projected.toFixed(2)} USD exceeds cap ${b.max_budget_usd.toFixed(2)} USD`,
            };
        }
    }

    // 3. scope (data classification)
    if (b.data_scope) {
        const raw = action.dataClassification;
        if (raw !== undefined) {
            const tag = raw.toLowerCase();
            const deny = b.data_scope.deny.map((s) => s.toLowerCase());
            const allow = b.data_scope.allow.map((s) => s.toLowerCase());
            if (deny.includes(tag)) {
                return {
                    kind: "deny",
                    check: "scope",
                    reason: `classification '${tag}' is on deny list`,
                };
            }
            if (allow.length > 0 && !allow.includes(tag)) {
                return {
                    kind: "deny",
                    check: "scope",
                    reason: `classification '${tag}' not in allow list ${JSON.stringify(allow)}`,
                };
            }
        }
    }

    // 4. rate_limit
    if (b.rate_limit) {
        const limit = b.rate_limit.requests_per_minute;
        const lower = ctx.nowEpoch - RATE_WINDOW_SECS;
        let count = 0;
        for (const t of ctx.recentCallTimestamps) {
            if (t > lower && t <= ctx.nowEpoch) count++;
        }
        if (count >= limit) {
            return {
                kind: "deny",
                check: "rate_limit",
                reason: `${count} calls in last 60s reached limit ${limit}`,
            };
        }
    }

    // 5. time_window
    if (b.time_window) {
        if (!inWindow(b.time_window.start, b.time_window.end, ctx.nowTzHhmm)) {
            return {
                kind: "deny",
                check: "time_window",
                reason: `current time ${ctx.nowTzHhmm} (${b.time_window.timezone}) outside window [${b.time_window.start}, ${b.time_window.end}]`,
            };
        }
    }

    // 6. signatures (M-of-N per role)
    if (b.required_signatures) {
        for (const req of b.required_signatures) {
            const got = action.signatures.filter((s) => s === req.role).length;
            if (got < req.threshold) {
                return {
                    kind: "deny",
                    check: "signatures",
                    reason: `role '${req.role}' has ${got} of ${req.threshold} required signatures`,
                };
            }
        }
    }

    // 7. delegation depth
    if (b.delegation) {
        if (action.delegationDepth > b.delegation.max_depth) {
            return {
                kind: "deny",
                check: "delegation_depth",
                reason: `delegation_depth = ${action.delegationDepth} exceeds max ${b.delegation.max_depth}`,
            };
        }
    }

    return { kind: "allow" };
}

/**
 * Compute `HH:MM` in the given IANA timezone from a unix-epoch second.
 *
 * Uses `Intl.DateTimeFormat` so no extra dep is needed. Falls back to
 * UTC if the runtime rejects the tz string.
 */
export function computeNowTzHhmm(epochSeconds: number, ianaTz: string): string {
    const date = new Date(epochSeconds * 1000);
    try {
        const fmt = new Intl.DateTimeFormat("en-GB", {
            timeZone: ianaTz,
            hour: "2-digit",
            minute: "2-digit",
            hour12: false,
        });
        // en-GB → "HH:MM"; some runtimes return "24:00" at midnight, normalise.
        const out = fmt.format(date).replace("24:", "00:");
        if (/^\d{2}:\d{2}$/.test(out)) return out;
        // Fall through to UTC fallback.
    } catch {
        // tz invalid; fall through.
    }
    const h = date.getUTCHours().toString().padStart(2, "0");
    const m = date.getUTCMinutes().toString().padStart(2, "0");
    return `${h}:${m}`;
}

/** `true` if `hhmm ∈ [start, end]`, handling wrap-around when start > end. */
function inWindow(start: string, end: string, hhmm: string): boolean {
    if (start <= end) {
        return hhmm >= start && hhmm <= end;
    }
    return hhmm >= start || hhmm <= end;
}

File: sdk/go/sauronid/evaluator.go
package sauronid

import (
	"fmt"
	"strings"
	"time"
)

// RateWindowSecs is the width of the rate-limit sliding window in seconds.
// Must match the Rust + TypeScript + Python evaluators.
const RateWindowSecs = 60

// Evaluate runs every applicable check from policy.Binding against action.
//
// It returns the first deny verdict, or {Kind: "allow"} if all checks pass.
//
// Order mirrors core::policy::compiler::compile:
//
//	allowlist -> budget -> scope -> rate_limit -> time_window ->
//	signatures -> delegation_depth.
//
// Reason strings are byte-equivalent with the TypeScript and Python SDKs
// so cross-impl tests can hard-code identical fixtures.
func Evaluate(policy *CompiledPolicy, action *Action, ctx *EvaluationContext) Verdict {
	b := policy.Binding

	// 1. allowlist (tool name)
	if b.AllowedTools != nil {
		found := false
		for _, t := range b.AllowedTools {
			if t == action.Tool {
				found = true
				break
			}
		}
		if !found {
			return Verdict{
				Kind:   "deny",
				Check:  "allowlist",
				Reason: fmt.Sprintf("tool '%s' not in allowlist", action.Tool),
			}
		}
	}

	// 2. budget
	if b.MaxBudgetUsd != nil {
		amount := 0.0
		if action.AmountUsd != nil {
			amount = *action.AmountUsd
		}
		projected := ctx.SpendTotalUsd + amount
		if projected > *b.MaxBudgetUsd {
			return Verdict{
				Kind:  "deny",
				Check: "budget",
				Reason: fmt.Sprintf(
					"projected spend %s USD exceeds cap %s USD",
					formatFixed2(projected),
					formatFixed2(*b.MaxBudgetUsd),
				),
			}
		}
	}

	// 3. scope (data classification)
	if b.DataScope != nil && action.DataClassification != nil {
		tag := strings.ToLower(*action.DataClassification)
		denyList := lowerAll(b.DataScope.Deny)
		allowList := lowerAll(b.DataScope.Allow)
		if contains(denyList, tag) {
			return Verdict{
				Kind:   "deny",
				Check:  "scope",
				Reason: fmt.Sprintf("classification '%s' is on deny list", tag),
			}
		}
		if len(allowList) > 0 && !contains(allowList, tag) {
			return Verdict{
				Kind:  "deny",
				Check: "scope",
				Reason: fmt.Sprintf(
					"classification '%s' not in allow list %s",
					tag,
					jsonArrayLiteral(allowList),
				),
			}
		}
	}

	// 4. rate_limit
	if b.RateLimit != nil {
		limit := b.RateLimit.RequestsPerMinute
		lower := ctx.NowEpoch - RateWindowSecs
		count := 0
		for _, t := range ctx.RecentCallTimestamps {
			if t > lower && t <= ctx.NowEpoch {
				count++
			}
		}
		if count >= limit {
			return Verdict{
				Kind:   "deny",
				Check:  "rate_limit",
				Reason: fmt.Sprintf("%d calls in last 60s reached limit %d", count, limit),
			}
		}
	}

	// 5. time_window
	if b.TimeWindow != nil {
		if !inWindow(b.TimeWindow.Start, b.TimeWindow.End, ctx.NowTzHhmm) {
			return Verdict{
				Kind:  "deny",
				Check: "time_window",
				Reason: fmt.Sprintf(
					"current time %s (%s) outside window [%s, %s]",
					ctx.NowTzHhmm,
					b.TimeWindow.Timezone,
					b.TimeWindow.Start,
					b.TimeWindow.End,
				),
			}
		}
	}

	// 6. signatures (M-of-N per role)
	if b.RequiredSignatures != nil {
		for _, req := range b.RequiredSignatures {
			got := 0
			for _, s := range action.Signatures {
				if s == req.Role {
					got++
				}
			}
			if got < req.Threshold {
				return Verdict{
					Kind:  "deny",
					Check: "signatures",
					Reason: fmt.Sprintf(
						"role '%s' has %d of %d required signatures",
						req.Role, got, req.Threshold,
					),
				}
			}
		}
	}

	// 7. delegation depth
	if b.Delegation != nil {
		if action.DelegationDepth > b.Delegation.MaxDepth {
			return Verdict{
				Kind:  "deny",
				Check: "delegation_depth",
				Reason: fmt.Sprintf(
					"delegation_depth = %d exceeds max %d",
					action.DelegationDepth, b.Delegation.MaxDepth,
				),
			}
		}
	}

	return Verdict{Kind: "allow"}
}

// ComputeNowTzHhmm returns "HH:MM" in the given IANA timezone for the
// supplied unix-epoch second.
//
// Falls back to UTC if time.LoadLocation rejects the timezone name.
func ComputeNowTzHhmm(epochSec int64, ianaTz string) string {
	t := time.Unix(epochSec, 0).UTC()
	loc, err := time.LoadLocation(ianaTz)
	if err == nil {
		local := t.In(loc)
		return local.Format("15:04")
	}
	return t.Format("15:04")
}

// inWindow returns true if hhmm is in [start, end] inclusive, with
// wrap-around support when start > end.
func inWindow(start, end, hhmm string) bool {
	if start <= end {
		return hhmm >= start && hhmm <= end
	}
	return hhmm >= start || hhmm <= end
}

func lowerAll(in []string) []string {
	out := make([]string, len(in))
	for i, s := range in {
		out[i] = strings.ToLower(s)
	}
	return out
}

func contains(list []string, x string) bool {
	for _, s := range list {
		if s == x {
			return true
		}
	}
	return false
}

// jsonArrayLiteral renders a string list as a compact JSON literal,
// matching TS JSON.stringify(["a","b"]) byte-for-byte: ["a","b"] with no
// inter-element whitespace.
func jsonArrayLiteral(items []string) string {
	var sb strings.Builder
	sb.WriteByte('[')
	for i, s := range items {
		if i > 0 {
			sb.WriteByte(',')
		}
		sb.WriteByte('"')
		sb.WriteString(s)
		sb.WriteByte('"')
	}
	sb.WriteByte(']')
	return sb.String()
}

// formatFixed2 mimics the JS Number.prototype.toFixed(2) output exactly:
// half-away-from-zero rounding to two decimals.
func formatFixed2(v float64) string {
	// JS toFixed uses IEEE round-half-to-even in some engines, but in
	// practice the test fixtures avoid the .005 boundary. Use Go default
	// "%.2f" which gives the same output for the values our reasons
	// produce.
	return fmt.Sprintf("%.2f", v)
}
